File: app/src/main/java/com/frank/ffmpeg/mp3/Mp3Lame.java
package com.frank.ffmpeg.mp3;

import com.frank.ffmpeg.AudioPlayer;

public class Mp3Lame {

    public Mp3Lame() {
        AudioPlayer.lameInitDefault();
    }

    Mp3Lame(Mp3LameBuilder builder) {
        initialize(builder);
    }

    private void initialize(Mp3LameBuilder builder) {
        AudioPlayer.lameInit(builder.inSampleRate, builder.outChannel, builder.outSampleRate,
                builder.outBitrate, builder.scaleInput, getIntForMode(builder.mode), getIntForVbrMode(builder.vbrMode), builder.quality, builder.vbrQuality, builder.abrMeanBitrate,
                builder.lowPassFreq, builder.highPassFreq, builder.id3tagTitle, builder.id3tagArtist,
                builder.id3tagAlbum, builder.id3tagYear, builder.id3tagComment);
    }

    public int encode(short[] buffer_l, short[] buffer_r,
                      int samples, byte[] mp3buf) {

        return AudioPlayer.lameEncode(buffer_l, buffer_r, samples, mp3buf);
    }

    int encodeBufferInterLeaved(short[] pcm, int samples,
                                byte[] mp3buf) {
        return AudioPlayer.encodeBufferInterleaved(pcm, samples, mp3buf);
    }

    public int flush(byte[] mp3buf) {
        return AudioPlayer.lameFlush(mp3buf);
    }

    public void close() {
        AudioPlayer.lameClose();
    }

    private static int getIntForMode(Mp3LameBuilder.Mode mode) {
        switch (mode) {
            case STEREO:
                return 0;
            case JSTEREO:
                return 1;
            case MONO:
                return 3;
            case DEFAULT:
                return 4;
        }
        return -1;
    }

    private static int getIntForVbrMode(Mp3LameBuilder.VbrMode mode) {
        switch (mode) {
            case VBR_OFF:
                return 0;
            case VBR_RH:
                return 2;
            case VBR_ABR:
                return 3;
            case VBR_MTRH:
                return 4;
            case VBR_DEFAUT:
                return 6;
        }
        return -1;
    }

}

File: app/src/main/java/com/frank/ffmpeg/mp3/Mp3Lame.kt
package com.frank.ffmpeg.mp3

import com.frank.ffmpeg.AudioPlayer

class Mp3Lame {

    constructor() {
        AudioPlayer.lameInitDefault()
    }

    internal constructor(builder: Mp3LameBuilder) {
        initialize(builder)
    }

    private fun initialize(builder: Mp3LameBuilder) {
        AudioPlayer.lameInit(builder.inSampleRate, builder.outChannel, builder.outSampleRate,
                builder.outBitrate, builder.scaleInput, getIntForMode(builder.mode), getIntForVbrMode(builder.vbrMode), builder.quality, builder.vbrQuality, builder.abrMeanBitrate,
                builder.lowPassFreq, builder.highPassFreq, builder.id3tagTitle, builder.id3tagArtist,
                builder.id3tagAlbum, builder.id3tagYear, builder.id3tagComment)
    }

    fun encode(buffer_l: ShortArray, buffer_r: ShortArray,
               samples: Int, mp3buf: ByteArray): Int {

        return AudioPlayer.lameEncode(buffer_l, buffer_r, samples, mp3buf)
    }

    internal fun encodeBufferInterLeaved(pcm: ShortArray, samples: Int,
                                         mp3buf: ByteArray): Int {
        return AudioPlayer.encodeBufferInterleaved(pcm, samples, mp3buf)
    }

    fun flush(mp3buf: ByteArray): Int {
        return AudioPlayer.lameFlush(mp3buf)
    }

    fun close() {
        AudioPlayer.lameClose()
    }

    private fun getIntForMode(mode: Mp3LameBuilder.Mode): Int {
        return when (mode) {
            Mp3LameBuilder.Mode.STEREO -> 0
            Mp3LameBuilder.Mode.JSTEREO -> 1
            Mp3LameBuilder.Mode.MONO -> 3
            Mp3LameBuilder.Mode.DEFAULT -> 4
        }
    }

    private fun getIntForVbrMode(mode: Mp3LameBuilder.VbrMode): Int {
        return when (mode) {
            Mp3LameBuilder.VbrMode.VBR_OFF -> 0
            Mp3LameBuilder.VbrMode.VBR_RH -> 2
            Mp3LameBuilder.VbrMode.VBR_ABR -> 3
            Mp3LameBuilder.VbrMode.VBR_MTRH -> 4
            Mp3LameBuilder.VbrMode.VBR_DEFAUT -> 6
        }
    }

}
